src/future/shared.rs

```diff
@@ -38,64 +39,76 @@ impl<F> fmt::Debug for Shared<F>
 {
     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
         fmt.debug_struct("Shared")
-            .field("id", &self.id)
             .field("inner", &self.inner)
             .finish()
     }
 }
 
 struct Inner<F: Future> {
-    next_clone_id: Mutex<u64>,
-    state: Mutex<State<F>>,
+    next_clone_id: AtomicUsize,
+    future: UnsafeCell<Spawn<F>>,
+    result: UnsafeCell<Option<Result<SharedItem<F::Item>, SharedError<F::Error>>>>,
+    unparker: Arc<Unparker>,
 }
 
-impl<F> fmt::Debug for Inner<F>
-    where F: Future + fmt::Debug,
-          F::Item: fmt::Debug,
-          F::Error: fmt::Debug,
-{
-    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        fmt.debug_struct("Inner")
-            .field("next_clone_id", &self.next_clone_id)
-            .field("state", &self.state)
-            .finish()
-    }
+struct Unparker {
+    state: AtomicUsize,
+    waiters: Mutex<HashMap<usize, Task>>,
```

Comment:
I know the performance probably won't matter for this use case, but `Mutex<HashMap<...>>` never sits well with me.

[**tangent**] If there were a concurrent hashmap in libstd, that'd be used here, but since there's not, and since futures-rs is such a fundamental crate, I assume a 3rd party crate won't be considered. I think it's worth noting because usually we think of crates.io as zero cost compared to a batteries-included std.

As a thought experiment, under what conditions could a crates.io concurrent map be considered? Would it need a certain amount of popularity/provenness? A 1.0 release? Some commitment to maintenance?